tests/build-std/main.rs

```diff
@@ -25,7 +25,6 @@ use std::path::Path;
 fn enable_build_std(e: &mut Execs, arg: Option<&str>) {
     e.env_remove("CARGO_HOME");
     e.env_remove("HOME");
-    e.arg("-Zno-index-update");
```

Comment:
This ended up being a bit of an odd coincidence, but what was happening is that this test relies on the index being updated by the Cargo used to drive the tests. The `cargo` run here is assumed to read that index and work with it correctly. If we ever change anything about how the index is checked out, then this breaks because the new `cargo` can't read the old index. 

That's basically what happened here where the new Cargo is requiring `refs/remotes/origin/HEAD` to resolve the most recent commit of the index, but previous Cargo would only fill in `refs/remotes/origin/master`. This meant that with `-Zno-index-update` the new `cargo` binary would fail to update the index and would think there were no crates.

I can't actually recall the original purpose for this argument myself, but it didn't seem to slow down too too much when adding this back in.

Reply:
Oh, that's a bit worrisome.  I think there are a fair number of users of `Zno-index-update`, and it looks like it just breaks with a pre-existing index.  Would it be possible to handle that? 

This was added from the original build-std tests where there were over 30 of them.  In #7350 the build-std suite was improved where only 3 "full" tests were moved to `build-std/main.rs`.

Reply:
I should have said, is there an *easy* way to handle it?  If not, I guess since it is an unstable feature, might not be worth putting in the effort if it isn't easy.

Reply:
Oh I think this is only localized to this test suite. If you were, for example, to do:

```
$ cargo +nightly test --no-run
$ rm -rf $HOME/.cargo/registry
$ CARGO_RUN_BUILD_STD_TESTS=1 rustup run nightly ./target/debug/deps/build_std-*
```

this breaks today. The tests rely on the index already being up-to-date, and this just happens to work because the Cargo we're testing and the outer cargo (which ran `cargo +nightly test`) agree on the index format, so the tests know they'll be able to use the index the outer Cargo populated.

We could perhaps handle this by fetching the `HEAD` reference into `refs/remotes/origin/master` and then reading that, but I'd just want to confirm that we're on the same page about the fallout we're dealing with.

Reply:
The fallout I'm thinking of, is if someone has the registry populated from a current version of Cargo, and then they update to a nightly with this PR, and the first command they run is `cargo build -Zno-index-update`, it will fail with `error: no matching package named …`, which is a bit confusing.

I'm not sure what the likelyhood is that someone runs with `-Zno-index-update` as the very first command after updating.  As I mentioned, only bother if it is easy to handle.  It seems like it is unlikely that it would be the first command someone would run.  I believe Crater does [an initial fetch](https://github.com/rust-lang/rustwide/blob/de7e3e71f95aefb39f0f715b60371aee4917f50e/src/workspace.rs#L326-L338), so crater should be safe I would think.

Reply:
It's true that it's a confusing error, yeah. Do you think it's worth trying to return a better error?

Reply:
Eh.  I'm inclined to say "no" because it is an unstable feature, and I think it is unlikely that someone will run that as the first command.

r=me if you think this is ready to go.